src/node_messaging.h

```diff
@@ -0,0 +1,250 @@
+typedef std::shared_ptr<ReferenceCountedSAB> ExternalSABReference;
+class ReferenceCountedSAB
+    : public std::enable_shared_from_this<ReferenceCountedSAB> {
+ public:
+  static ExternalSABReference ForIncomingSharedArrayBuffer(
+      Environment* env, v8::Local<v8::Context> context,
+      v8::Local<v8::SharedArrayBuffer> source);
+  ~ReferenceCountedSAB();
+
+  // Create a SharedArrayBuffer object for a specific Environment and Context.
+  // The created SharedArrayBuffer will be in externalized mode and has
+  // a hidden object attached to it, during whose lifetime the reference
+  // count if increased by 1.
+  v8::MaybeLocal<v8::SharedArrayBuffer> GetSharedArrayBuffer(
+      Environment* env, v8::Local<v8::Context> context);
+
+  ReferenceCountedSAB(ReferenceCountedSAB&& other) = delete;
+  ReferenceCountedSAB& operator=(ReferenceCountedSAB&& other) = delete;
+  ReferenceCountedSAB& operator=(const ReferenceCountedSAB&) = delete;
+  ReferenceCountedSAB(const ReferenceCountedSAB&) = delete;
+
+ private:
+  explicit ReferenceCountedSAB(void* data, size_t size);
+
+  // Attach a lifetime tracker object with a reference count to `target`.
+  v8::Maybe<bool> AssignToSharedArrayBuffer(
+      Environment* env,
+      v8::Local<v8::Context> context,
+      v8::Local<v8::SharedArrayBuffer> target);
+
+  void* data = nullptr;
+  size_t size = 0;
+};
+
+
+// Represents a single communication message. The only non-standard extension
+// here is passing of a separate flag that the Workers implementation uses
+// for internal cross-thread information passing.
+class Message {
+ public:
+  explicit Message(int32_t flag = MESSAGE_FLAG_NONE);
+  Message(void* buf, size_t length, int32_t flag);
+  ~Message();
+
+  Message(Message&& other);
+  Message& operator=(Message&& other);
+  Message& operator=(const Message&) = delete;
+  Message(const Message&) = delete;
+
+  // Deserialize the contained JS value. May onle be called once, and only
+  // after Serialize() has been called (e.g. by another thread).
+  v8::MaybeLocal<v8::Value> Deserialize(Environment* env,
+                                        v8::Local<v8::Context> context);
+  // Serialize a JS value, and optionally transfer objects, into this message.
+  // The Message object retains ownership of all transferred objects until
+  // deserialization.
+  v8::Maybe<bool> Serialize(Environment* env,
+                            v8::Local<v8::Context> context,
+                            v8::Local<v8::Value> input,
+                            v8::Local<v8::Value> transfer_list);
+
+  // Internal method of Message that is called when a new SharedArrayBuffer
+  // object is encountered in the incoming value's structure.
+  void AddSharedArrayBuffer(ExternalSABReference ref);
+  // Internal method of Message that is called once serialization finishes
+  // and that transfers ownership of `data` to this message.
+  void AddMessagePort(std::unique_ptr<MessagePortData>&& data);
```

Comment:
Make these private?

Reply:
I think that would require giving the `SerializerDelegate` that accesses these a publicly visibly identifier, right?